### DISCUSSION TITLE
tRPC error on redirect from NextJs middleware

### DISCUSSION THREAD

**Discussion by stychu** [trpc/trpc#3075] [Q&A]
Hello, Im trying to implement global redirect for unauthenticated users but Im having problems with tRPC api routes whenever user access token is invalid. NextJs middleware picks up tRPC api route which expect json response but middleware redirects to the signin page thus crashing tRPC and not redirecting to the page. Is there a way I could overcome this via middleware or do I need to omit trpc api from middleware and handle it client side?? I would love if it possible to handle this somehow in nextjs middleware.

----
<img width="1215" alt="Screenshot 2022-10-25 at 01 30 31" src="https://user-images.githubusercontent.com/9383657/197771325-73bea602-1f80-4794-8572-5a2f86e1c11e.png">
<img width="547" alt="Screenshot 2022-10-25 at 01 30 47" src="https://user-images.githubusercontent.com/9383657/197771336-fac36e3c-33ae-43a3-8dc3-871e339adbfe.png">

I appreciate any comments! Thanks

**Comment by fusion4github** [comment]
facing the same issue :/

**Reply by stychu** [reply]
@fusion4github Do you use custom middleware or maybe trying with nextAuth middleware??

Anyway here are my findings https://github.com/trpc/trpc/issues/3081

**Comment by chrisflatley** [comment]
I had the same when fiddling with SSG Helpers and SSR. 

Since trpc protects itself, as do my other /apis/ I dropped an exception in the middleware matcher

```
import { withAuth } from "next-auth/middleware";

export const config = {
  matcher: ["/", "/((?!api|favicon.ico).*)"],
};
export default withAuth({});
```


The matchers (copied from somewhere, sorry I don't remember where):

- Everything not under /api or /favicon.ico requires auth. 
- I had to put the / explicitly to match the home

**Reply by stychu** [reply]
Yes. Obviously this work as you exclude `/api` but I want `api` to be included in the middleware. Until tRPC implements the thrown error for redirect from its query call I also do exclude `/api/trpc` from middleware checks
